# Introduction and Setup
{:.no_toc}

* TOC
{:toc}

# Introduction

## General Information

Thank you for your attendance & welcome to the *Introduction to Bash: Using the Terminal For Bioinformatics* Workshop.
This is an offering by the University of Adelaide, Bioinformatics Hub which is a centrally funded initiative from the Department of Vice-Chancellor (Research), with the aim of assisting & enabling researchers in their work.
Training workshops & seminars such as this one are an important part of this initiative.

Some additional resources run by the Bioinformatics Hub which may be of interest beyond today are:

- A University web-page at http://www.adelaide.edu.au/bioinformatics-hub/
- To be kept up to date on upcoming events and workshops, please join the internal Bioinformatics mailing list on http://list.adelaide.edu.au/mailman/listinfo/ bioinfo
- A Twitter account https://twitter.com/UofABioinfoHub/
- An active Slack team for discussing Bioinformatics questions with the local community. Slack teams do require an invitation to join, so please email the Hub on bioinf_hub@adelaide.edu.au to join the community. All are welcome.

Today’s workshop has been put together based on previous material and courses prepared by Dr Stephen Bent (University of Queensland), with generous technical support & advice provided by Dr Nathan Watson-Haigh and Dr Dan Kortschak.
We hope it will be useful in enabling you to continue and to advance your research.

## Course Summary

Next-generation sequencing (NGS) has become an important tool in assessing biological signal within an organism or population. Stemming from previous technologies that were costly and time-consuming to run, NGS platforms are relatively cheap and enable the investigation of the genome, transcriptome, methylome etc at extremely high resolution. The high-throughput of these machines also has unique challenges, and it is important that scientists are aware of the potential limitations of the platforms and the issues involved with the production of good quality data.

In this course, we will illustrate the fundamentals of Illumina NGS technology (the current market leader in the production of sequencing data), describe the affect that library preparation can have on downstream data, as well as running through a basic workflow used to analyse NGS data.

## Today's Instructors

Today your helpers will be

- Steve Pederson, Hien To, Alastair Ludington (Bioinformatics Hub)
- Jimmy Breen (Robinson Research Institute & Bioinformatics Hub)
- Robert Qiao, Exe Escobedo (Phoenix Support Team)
- Terry Bertozzi (SA Museum)
- Aaron Casey (SAHMRI)

## Recommendations For Working

In the following pages, we strongly encourage you to manually type all commands.
The mistakes you will inevitably make will actually be important learning steps.
Additionally, in your work beyond today, you will probably not have any instructions to follow.
The experience of typing these commands will equip you for future work far better than if you simply copy & paste.

For today’s session, you will also be provided with red post-it notes.
**Please use these to signal whether you need help or not by placing them on your monitors**.
These are easy for instructors to spot so we can make our way over, although do be aware that there will be times when all instructors are busy.
This will be important as we all set our computers up in the following section.

## Computer Setup

Previously we have run these sessions using Virtual Machines, but today we are opting for running all sessions using the tools locally installed on your own machines today.
*As there will be a huge variety of laptops in the room, the initial set-up may be a difficult to begin with*, but will settle as the day progresses.
Although this will present numerous challenges from our perspective, we feel this is a better approach as everyone will leave knowing exactly how to get the ball rolling on their own familiar systems.

Instructions for how to set-up everything on your own machine are at the following pages.
Please follow these very carefully, and call an instructor over if you have trouble.
In particular, the Windows set-ups may take a while to get sorted.

- [Mac/OSX](../install/osxInstall)
- [Windows](../install/windowsInstall)

If you are already running Ubuntu, your computer will already be set-up correctly.

[Home](../)
